bundles/org.openhab.binding.sony/src/main/java/org/openhab/binding/sony/internal/providers/sources/SonyFolderSource.java

```diff
@@ -0,0 +1,215 @@
+import org.openhab.binding.sony.internal.SonyUtil;
+import org.openhab.binding.sony.internal.providers.models.SonyDeviceCapability;
+import org.openhab.binding.sony.internal.providers.models.SonyThingDefinition;
+
+/**
+ * An implementation of a {@link SonySource} that will source thing types from
+ * json files within the user data folder
+ *
+ * @author Tim Roberts - Initial contribution
+ */
+@NonNullByDefault
+public class SonyFolderSource extends AbstractSonySource {
+    /** The base folder to use */
+    private static final String FOLDERBASE = ConfigConstants.getUserDataFolder() + File.separator + "sony";
+
+    /** Various property keys used by this source */
+    private static final String PROP_WATCHDOG_INTERVAL = "SonyFolderSource.WatchDog.Interval";
+    private static final String PROP_THINGTYPES = "SonyFolderSource.Folder.ThingTypes";
+    private static final String PROP_DEFTYPES = "SonyFolderSource.Folder.DefinitionTypes";
+    private static final String PROP_DEFCAPS = "SonyFolderSource.Folder.DefinitionCapabilities";
+
+    /** Folder paths */
+    private final String folderThingTypes;
+    private final String folderDefThingTypes;
+    private final String folderDefCapability;
+
+    /** The folder watcher (null if none being watched) */
+    private final AtomicReference<@Nullable Future<?>> watcher = new AtomicReference<>(null);
+
+    /** The folder watch dog (null if none being watched) */
+    private final AtomicReference<@Nullable Future<?>> watchDog = new AtomicReference<>(null);
+
+    /**
+     * Constructs the source and starts the various threads
+     * 
+     * @param scheduler a non-null scheduler to use
+     * @param properties a non-null, possibly empty map of properties
+     */
+    public SonyFolderSource(final ScheduledExecutorService scheduler, final Map<String, String> properties) {
+        Objects.requireNonNull(scheduler, "scheduler cannot be null");
+        Objects.requireNonNull(properties, "properties cannot be null");
+
+        // create the folders we will use
+        folderThingTypes = FOLDERBASE + getProperty(properties, PROP_THINGTYPES);
+        folderDefThingTypes = FOLDERBASE + getProperty(properties, PROP_DEFTYPES);
+        folderDefCapability = FOLDERBASE + getProperty(properties, PROP_DEFCAPS);
+
+        createFolder(folderThingTypes);
+        createFolder(folderDefThingTypes);
+        createFolder(folderDefCapability);
+
+        // Setup our watcher
+        watcher.getAndSet(scheduler.submit(new Watcher()));
+
+        // Keeps the watcher alive in case it encounters an error
+        final int watchDogTime = getPropertyInt(properties, PROP_WATCHDOG_INTERVAL);
+        SonyUtil.cancel(watchDog.getAndSet(scheduler.schedule(() -> {
+            final Future<?> fut = watcher.get();
+            if (fut == null || fut.isDone()) {
+                SonyUtil.cancel(watcher.getAndSet(scheduler.submit(new Watcher())));
+            }
+        }, watchDogTime, TimeUnit.SECONDS)));
```

Comment:
Actually it does run more than once (get's first, then sets the schedule).  However - there is absolutely no need for that SonyUtil.cancel since this is the constructor and the watchdog hasn't been set yet.  (both it and the watcher.getAndSet can be made to just .set)

Reply:
Well the `schedule` method just delays the execution of something. `scheduleAtFixedRate` and `scheduleWithFixedDelay` are the ones that do periodic execution. That is what I was trying to point out since your comments imply that it should be run periodically.